kafkaconsumer/partition_manager.go

```diff
@@ -0,0 +1,286 @@
+	}
+}
+
+// closePartitionOffsetManager stops the partition offset manager for this partitions, and will write
+// any error that may happen to the logger.
+func (pm *partitionManager) closePartitionOffsetManager() {
+	if err := pm.offsetManager.Close(); err != nil {
+		pm.logf("Failed to close partition offset manager: %s\n", err)
+	}
+}
+
+// startPartitionConsumer starts a sarama consumer for the partition under management.
+// This function will retry any error that may occur. The error return value is nil once
+// it successfully has started the partition consumer, or tomb.ErrDying if the partition
+// manager was interrupted. Any other error is not recoverable.
+func (pm *partitionManager) startPartitionConsumer(initialOffset int64) (sarama.PartitionConsumer, error) {
+	var (
+		pc  sarama.PartitionConsumer
+		err error
+	)
+
+	for {
+		pc, err = pm.parent.consumer.ConsumePartition(pm.partition.Topic().Name, pm.partition.ID, initialOffset)
+		switch err {
+		case nil:
+			switch initialOffset {
+			case sarama.OffsetNewest:
+				pm.logf("Started consumer for new messages only.")
+			case sarama.OffsetOldest:
+				pm.logf("Started consumer at the oldest available offset.")
+			default:
+				pm.logf("Started consumer at offset %d.", initialOffset)
+			}
+
+			// We have a valid partition consumer so we can return
+			return pc, nil
+
+		case sarama.ErrOffsetOutOfRange:
+			// The offset we had on file is too old. Restart with initial offset
+			if pm.parent.config.Offsets.Initial == sarama.OffsetNewest {
+				pm.logf("Offset %d is no longer available. Trying again with new messages only...", initialOffset)
+			} else if pm.parent.config.Offsets.Initial == sarama.OffsetOldest {
+				pm.logf("Offset %d is no longer available. Trying again with he oldest available offset...", initialOffset)
+			}
+			initialOffset = pm.parent.config.Offsets.Initial
+
+			continue
+
+		default:
+			// Assume the problem is temporary; just try again.
+			pm.logf("Failed to start consuming partition: %s. Trying again in 1 second...\n", err)
+			select {
+			case <-pm.t.Dying():
+				return nil, tomb.ErrDying
+			case <-time.After(1 * time.Second):
+				continue
+			}
+		}
+
+	}
+}
+
+// closePartitionConsumer closes the sarama consumer for the partition under management.
+func (pm *partitionManager) closePartitionConsumer(pc sarama.PartitionConsumer) {
+	if err := pc.Close(); err != nil {
+		pm.logf("Failed to close partition consumer: %s\n", err)
+	}
+}
+
+// waitForProcessing waits for all the messages that were consumed for this partition to be processed.
+// The processing can take at most MaxProcessingTime time. After that, those messages are consisered
+// lost and will not be committed. Note that this may cause messages to be processed twice if another
+// partition consumer resumes consuming from this partition later.
+func (pm *partitionManager) waitForProcessing() {
+	nextOffset, _ := pm.offsetManager.NextOffset()
+	lastProcessedOffset := nextOffset - 1
```

Comment:
seems reasonable to me... early returns would make it much prettier though, e.g. 

```if lastConsumedOffset == -1 {
  return
}
// ...
```